src/common/src/test/scala/common/APISpec.scala

```diff
@@ -0,0 +1,434 @@
+
+      val dfNds = way1.getSeq[Row](way1.fieldIndex("nds")).map(_.getAs[Long]("ref"))
+
+      assert(dfNds === way2.nds)
+    }
+
+    it("should pass relation members through") {
+      val conditions = 'id === 83689 and 'version === 1
+
+      val relation1 = df.where('type === "relation" and conditions).first()
+      val relation2 = ds.where('type === RelationType and conditions).first()
+
+      val dfMembers = relation1
+        .getSeq[Row](relation1.fieldIndex("members"))
+        .map(row =>
+          (row.getAs[String]("type"): Byte, row.getAs[Long]("ref"), row.getAs[String]("role")))
+
+      val historyMembers = relation2.members.map(m => (m.`type`, m.ref, m.role))
+
+      assert(dfMembers === historyMembers)
+    }
+  }
+
+  describe("asSnapshot") {
+    val df = SnapshotDF
+    val snapshot = asSnapshot(df)
+    val ds = snapshot.dataset
+
+    it("should contain the same fields") {
+      assert(df.schema.fieldNames === ds.schema.fieldNames)
+    }
+
+    it("should fail with partial data") {
+      assertThrows[AnalysisException] {
+        asSnapshot(df.select('id, 'type, 'tags, 'lat, 'lon))
+      }
+    }
+
+    it("should include additional columns") {
+      val ds = asHistory(df.withColumn("something", lit("something")))
+
+      assert(ds.schema.fieldNames.contains("something"))
+
+      val row = ds.select('something).first()
+
+      assert(row.getAs[String]("something") === "something")
+    }
+
+    it("should coerce `type` to a Byte") {
+      val typeType = ds.schema.fields(ds.schema.fieldIndex("type"))
+      assert(typeType.dataType.isInstanceOf[ByteType])
+    }
+
+    it("should compress relation members") {
+      import osmesa.common.implicits._
+
+      val el = ds.relations.where('id === 58446 and 'version === 150).first()
+
+      assert(el.members.head.`type` === NodeType)
+    }
+  }
+
+  describe("Dataset[OSM] with History") {
+    val history = asHistory(HistoryDF)
+    val timestamp = Timestamp.valueOf("2012-01-01 00:00:00")
+
+    import implicits._
+
+    describe("nodes") {
+      it("should produce only nodes") {
+        assert(history.nodes.count === 290861)
+      }
+
+      it("should include Tags") {
+        assert(history.nodes.schema.fieldNames.contains("tags"))
+      }
+    }
+
+    describe("ways") {
+      it("should produce only ways") {
+        assert(history.ways.count === 75849)
+      }
+    }
+
+    describe("relations") {
+      it("should produce only relations") {
+        assert(history.relations.count === 4825)
+      }
+    }
+
+    describe("snapshot") {
+      val snapshot = history.snapshot(timestamp)
+      val ds = snapshot.dataset
+      ds.cache
+
+      it("should store the timestamp") {
+        assert(snapshot.timestamp === timestamp)
+      }
+
+      it("should contain a snapshot of data") {
+        assert(ds.where('type === NodeType).count === 29012)
+      }
+    }
+
+    describe("withValidity") {
```

Comment:
Is withValidity / Way Validity documented anywhere?

Reply:
Not yet. I've only started sketching out the scaladoc.

The implementation is here: https://github.com/azavea/osmesa/pull/103/files#diff-959969bab4330f1cdea4195279fe192eR445

(it applies a window function over each element in order to add `validUntil`)